# On the importance of data transformation for data integration in single-cell RNA sequencing analysis

This repository is for the investigation of the impact of data normalization in integrative analysis of the single-cell RNA-seq data.

Check out this [document](https://iron-lion.github.io/scRNAseq-preprocessing-impact-doc/chapters/intro/intro.html).

----
## More information
* [BMC Bioinformatics](https://doi.org/10.1186/s12859-024-05788-5)
* [bioRxiv](https://doi.org/10.1101/2022.07.19.500522)
* [Poster](https://github.com/iron-lion/scRNAseq-preprocessing-impact/blob/main/figures/06_22_poster.png)


----
## Reference
 - Luecken, M.D., Büttner, M., Chaichoompu, K. et al. [Benchmarking atlas-level data integration in single-cell genomics](https://doi.org/10.1038/s41592-021-01336-8). Nat Methods 19, 41–50 (2022). 
 - Constantin Ahlmann-Eltze, Wolfgang Huber. [Comparison of Transformations for Single-Cell RNA-Seq Data](https://doi.org/10.1101/2021.06.24.449781) bioRxiv (2021).
 - Tran, H.T.N., Ang, K.S., Chevrier, M. et al. [A benchmark of batch-effect correction methods for single-cell RNA sequencing data](https://doi.org/10.1186/s13059-019-1850-9). Genome Biol 21, 12 (2020).
 - Wang, C., Gao, X. & Liu, J. [Impact of data preprocessing on cell-type clustering based on single-cell RNA-seq data](https://doi.org/10.1186/s12859-020-03797-8). BMC Bioinformatics 21, 440 (2020). 
 - Tian, L., Dong, X., Freytag, S. et al. [Benchmarking single cell RNA-sequencing analysis pipelines using mixture control experiments](https://doi.org/10.1038/s41592-019-0425-8). Nat Methods 16, 479–487 (2019). 
